File: csi/java/src/main/java/org/shypl/biser/csi/server/ConnectionProcessorMessaging.java
package org.shypl.biser.csi.server;

import org.shypl.biser.csi.ByteBuffer;
import org.shypl.biser.csi.Protocol;
import org.shypl.biser.csi.ProtocolException;

class ConnectionProcessorMessaging extends ConnectionProcessor {
	private final AbstractClient client;
	
	private State      state  = State.FLAG;
	private ByteBuffer buffer = new ByteBuffer();
	private int messageId;
	private int messageSize;
	
	public ConnectionProcessorMessaging(AbstractClient client) {
		super();
		this.client = client;
	}
	
	@Override
	public void processData() throws ProtocolException {
		do {
			switch (state) {
				case FLAG:
					readFlag();
					break;
				case MESSAGE_ID:
					readMessageId();
					break;
				case MESSAGE_SIZE:
					readMessageSize();
					break;
				case MESSAGE_BODY:
					readMessageBody();
					break;
				case OUTGOING_MESSAGE_RECEIVED:
					readOutgoingMessageReceived();
					break;
			}
		}
		while (connection.isReadable());
	}
	
	@Override
	public void processClose(boolean breaking) {
		if (breaking) {
			client.handleConnectionBreaking();
		}
		else {
			client.handleConnectionClosed();
		}
	}
	
	private void readFlag() throws ProtocolException {
		byte flag = connection.read();
		switch (flag) {
			case Protocol.PING:
				connection.syncSend(Protocol.PING);
				client.sendLastMessageReceivedFlag();
				break;
			
			case Protocol.MESSAGE:
				prepareMessage();
				break;
			
			case Protocol.MESSAGE_RECEIVED:
				prepareOutgoingMessageReceived();
				break;
			
			case Protocol.CLOSE:
				connection.syncClose();
				break;
			
			default:
				throw new ProtocolException(String.format("Messaging: Invalid flag  0x%s (%s)", Integer.toString(flag & 0xFF, 16), (char)(flag & 0xFF)));
		}
	}
	
	private void prepareOutgoingMessageReceived() {
		state = State.OUTGOING_MESSAGE_RECEIVED;
		buffer.clear();
	}
	
	private void readOutgoingMessageReceived() {
		connection.read(buffer, 4 - buffer.getReadableBytesLength());
		if (4 == buffer.getReadableBytesLength()) {
			state = State.FLAG;
			messageId = buffer.readInt();
			buffer.clear();
			client.processOutgoingMessageReceived(messageId);
		}
	}
	
	private void prepareMessage() {
		state = State.MESSAGE_ID;
		buffer.clear();
	}
	
	private void readMessageId() {
		connection.read(buffer, 4 - buffer.getReadableBytesLength());
		if (4 == buffer.getReadableBytesLength()) {
			state = State.MESSAGE_SIZE;
			messageId = buffer.readInt();
			buffer.clear();
		}
	}
	
	private void readMessageSize() {
		connection.read(buffer, 4 - buffer.getReadableBytesLength());
		if (4 == buffer.getReadableBytesLength()) {
			state = State.MESSAGE_BODY;
			messageSize = buffer.readInt();
			buffer.clear();
		}
	}
	
	private void readMessageBody() {
		connection.read(buffer, messageSize - buffer.getReadableBytesLength());
		if (messageSize == buffer.getReadableBytesLength()) {
			state = State.FLAG;
			client.receiveMessage(messageId, buffer.readBytes());
		}
	}
	
	private enum State {
		FLAG, MESSAGE_ID, MESSAGE_SIZE, MESSAGE_BODY, OUTGOING_MESSAGE_RECEIVED
	}
}

File: kotlin/biser-kotlin-js/src/main/kotlin/org/shypl/biser/csi/client/ConnectionProcessorMessaging.kt
package org.shypl.biser.csi.client

import org.shypl.biser.csi.ConnectionCloseReason
import org.shypl.biser.csi.Protocol
import org.shypl.biser.io.ByteArrayInputData
import org.shypl.biser.io.InputData
import kotlin.browser.window

internal class ConnectionProcessorMessaging : ConnectionProcessor() {
	private val STATE_MESSAGE_ID = 0
	private val STATE_MESSAGE_SIZE = 1
	private val STATE_MESSAGE_BODY = 2
	private val STATE_OUTGOING_MESSAGE_RECEIVED = 3
	
	private var activity = false
	private var activityChecker = 0
	
	private var state = STATE_MESSAGE_ID
	private var messageId = 0
	
	override fun init(connection: Connection) {
		super.init(connection)
		activity = true
		activityChecker = window.setInterval(::checkActivity, connection.activityTimeout * 1000 / 2)
	}
	
	override fun destroy() {
		super.destroy()
		window.clearInterval(activityChecker)
	}
	
	override fun processClose() {
		if (connection.recoveryTimeout <= 0) {
			connection.logger.debug("Messaging: Interrupted, close")
			connection.close(ConnectionCloseReason.RECOVERY_REJECT)
		} else {
			connection.logger.debug("Messaging: Interrupted, switch to Recovery")
			connection.setProcessor(ConnectionProcessorRecovery())
		}
	}
	
	override fun processData(data: InputData) {
		activity = true
		do {
			super.processData(data)
		} while (isCanReadData(data))
	}
	
	override fun processDataFlag(flag: Byte) {
		when (flag) {
			Protocol.PING             -> Unit
			Protocol.MESSAGE          -> prepareMessage()
			Protocol.MESSAGE_RECEIVED -> prepareOutgoingMessageReceived()
			else                      -> super.processDataFlag(flag)
		}
		
	}
	
	override fun processDataBody(data: InputData) {
		when(state) {
			STATE_MESSAGE_ID -> {
				messageId = data.readInt()
				state = STATE_MESSAGE_SIZE
				setDataExpectBody(4)
			}
			STATE_MESSAGE_SIZE -> {
				state = STATE_MESSAGE_BODY
				setDataExpectBody(data.readInt())
			}
			STATE_MESSAGE_BODY -> {
				val message: ByteArray = ByteArray(data.readableBytes)
				data.readBytes(message)
				connection.receiveMessage(messageId, ByteArrayInputData(message))
			}
			STATE_OUTGOING_MESSAGE_RECEIVED -> {
				connection.processOutgoingMessageReceived(data.readInt())
			}
		}
	}
	
	private fun prepareOutgoingMessageReceived() {
		state = STATE_OUTGOING_MESSAGE_RECEIVED
		setDataExpectBody(4)
	}
	
	private fun prepareMessage() {
		state = STATE_MESSAGE_ID
		setDataExpectBody(4)
	}
	
	private fun checkActivity() {
		if (activity) {
			activity = false
			connection.sendPing()
		} else {
			window.clearInterval(activityChecker)
			connection.logger.warn("Messaging: Connection is not active, interrupt")
			connection.interrupt()
		}
	}
}
